Unify separation by implicitly including algorithm identifiers

blipp opened the pull request:
I would like to make an alternative proposition to #126, based on it.

In this proposition, `identifier` is used as implicit parameter to `LabeledExtract` and `LabeledExpand`. This can be justified because the document already assumes that `kem_id`, `kdf_id`, `aead_id` are implicitly encoded into the implementation. The advantage of having `identifier` implicit is that this removes clutter from the other function definitions; this might make implementation work easier because the calls to LabeledExtract and LabeledExpand then have fewer nested parameters.

On the question if `mode` should be in the identifier: if `mode` should be in the identifier, it is hard to let `identifier` be an implicit parameter, because the `mode` parameter is an explicit parameter (which is needed for `VerifyPSKInputs`). I thought about if we can find a reason why *not* including `mode` in the identifier is justified from a provable security point-of-view. I came to the conclusion this works out, because:

The first calls in `KeySchedule` are, from a variable dependencies point of view:
```
  psk_hash = LabeledExtract(zero(0), "psk_hash", psk)
  secret = LabeledExtract(psk_hash, "secret", zz)
```

All the following KDF calls include `mode` because it is in the `key_schedule_context`, so those are safe. Values of `secret` cannot collide between different modes, because:
- `zz` is the same for the Base and PSK modes, but `psk_hash` is different for these modes (default_psk vs non-defaut)
- `zz` is the same for the Auth and AuthPSK modes, but `psk_hash` is different for these modes (default_psk vs non-defaut)

This means we do not need to domain-separate by the mode in each call, because the two LabeledExtracts above already have a clear-enough dependency on the mode; which allows us to have `identifier` be implicit and have cleaner function definitions.

chris-wood commented on draft-irtf-cfrg-hpke.md:
```diff
@@ -433,6 +440,15 @@ def AuthDecap(enc, skR, pkS):
+identifier = I2OSP(kem_id, 2)
+~~~
+
+
```
```suggestion
```

bifurcation commented on draft-irtf-cfrg-hpke.md:
```diff
@@ -347,21 +347,27 @@ A set of algorithm identifiers for concrete instantiations of these
+implicit from the implementation and not passed as parameter. If used
+inside a KEM algorithm, `identifier` MUST identify this KEM algorithm;
+if used in the remainder of HPKE, it MUST identify the entire ciphersuite
+in use. See sections {{dhkem}} and {{encryption-context}} for details.
```
This seems likely to be confused.  Would it be clearer to add the identifier as a parameter to these methods, and explicitly pass in the right one?

bifurcation replied:
Note that the suite ID is a different length than the individual KEM ID.  Do we need to worry about collision with the label?

blipp replied:
> Note that the suite ID is a different length than the individual KEM ID. Do we need to worry about collision with the label?

Uh, good one. Yes, theoretically `aead_id` could collide with the labels `eae_prk` and `zz`. Not for the AEADs in the current document though: `hex(ord("z"))` is `0x7a` and `hex(ord("e"))` is `0x65`, and all current AEADs have 0s in the first byte.

We could have the `suite_id` for the KEMs be `concat(I2OSP(kem_id, 2), zero(4))` to avoid that forever.

bifurcation commented on draft-irtf-cfrg-hpke.md:
```diff
@@ -551,8 +566,18 @@ then the recipient is assured that the sender held the corresponding
+`LabeledExpand` is defined based on them as follows:
+
+~~~
+identifier = concat(
```
`suite_id`  for consistency?

blipp commented:
- I resolved the possible collision between the KEM KDF calls and the remainder of HPKE by adding `KEM` and `DEM` after `RFCXXXX `. I feel like this was overdue anyway.
- I changed `identifier` to `suite_id`.
- Regarding @bifurcation's comment
> This seems likely to be confused. Would it be clearer to add the identifier as a parameter to these methods, and explicitly pass in the right one?

Hm, I am torn on this one. These two possibly *different* KDFs that are both referenced with the *same names* `LabeledExtract` and `LabeledExpand` have been in the document since we added a KDF in the KEM. Adding `suite_id` just makes this possible confusion problem more apparent I guess. The test vectors will catch it at the end, but if we can make the confusion less likely, it would be nice for the developers.

I am fond of the idea to leave `suite_id` implicit, because making it explicit will clutter the calls to `LabeledExtract` and `LabeledExpand`. What else could we do? Separate names like `LabeledExtractKEM` and `LabeledExtractDEM`?

chris-wood commented:
> I am fond of the idea to leave suite_id implicit, because making it explicit will clutter the calls to LabeledExtract and LabeledExpand. What else could we do? Separate names like LabeledExtractKEM and LabeledExtractDEM?

I'm not sure I see much difference between a separate name for the DHKEM LabeledExtract/Expand functions, and passing the suite_id in as a parameter to the existing LabeledExtract/Expand functions. On balance, I think I too prefer what's currently written (implicit parameters).

bifurcation commented on draft-irtf-cfrg-hpke.md:
```diff
@@ -347,21 +347,28 @@ A set of algorithm identifiers for concrete instantiations of these
+implicit from the implementation and not passed as parameter. If used
+inside a KEM algorithm, `suite_id` MUST start with "KEM" and identify
+this KEM algorithm; if used in the remainder of HPKE, it MUST start with
+"DEM" and identify the entire ciphersuite in use. See sections {{dhkem}}
```
Nit: I would use "HPKE" for the non-KEM thing.

blipp replied:
Good point – I just made this change.
